### PR TITLE
Add varchar/char cast pushdown support in Redshift

### PR DESCRIPTION
**Pull request by mayankvadariya**
## Description

This PR adds support for cast projection pushdown for the Redshift connector for char/varchar types. 

- char -> char
- varchar -> char, varchar

Additionally, enforces case-sensitive comparison by using Redshift COLLATE function https://docs.aws.amazon.com/redshift/latest/dg/r_COLLATE.html

This PR depends on https://github.com/trinodb/trino/pull/22951

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Section
* Add varchar/char cast pushdown support in Redshift. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementStringLpadFunction.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        Constant argument2 = captures.get(ARGUMENT2);
+        Constant argument3 = captures.get(ARGUMENT3);
+
+        JdbcTypeHandle typeHandle = ((JdbcColumnHandle) context.getAssignment(argument1.getName())).getJdbcTypeHandle();
+
+        ParameterizedExpression translatedArgument1 = context.rewriteExpression(argument1).orElseThrow();
+        ParameterizedExpression translatedArgument2 = context.rewriteExpression(argument2).orElseThrow();
+        ParameterizedExpression translatedArgument3 = context.rewriteExpression(argument3).orElseThrow();
+
+        JdbcTypeHandle updatedTypeHandle = new JdbcTypeHandle(typeHandle.jdbcType(), typeHandle.jdbcTypeName(), Optional.empty(), typeHandle.decimalDigits(), typeHandle.arrayDimensions(), typeHandle.caseSensitivity());
```
Comment (SemionPar) [inline plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementStringLpadFunction.java:80]: *Question*: I guess this change 
https://github.com/trinodb/trino/pull/23490/files#diff-e7f2966e1e2e070e8643b5563c7ba62f7dc0652f76fbf446ef389b75ef4883fbL653-R672
is to accommodate unbounded size here?

I wonder, what if we used the size from the `lpad(1param, [size], 3param)` instead?

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+    @Override
+    protected List<CastTestCase> failCast()
+    {
+        return ImmutableList.of(
+                new CastTestCase("c_varchar_decimal", "smallint"),
+                new CastTestCase("c_varchar_decimal_sign", "smallint"),
+                new CastTestCase("c_varchar_alpha_numeric", "smallint"),
+                new CastTestCase("c_char_50", "smallint"),
+                new CastTestCase("c_char_numeric", "smallint"),
+                new CastTestCase("c_bpchar_numeric", "smallint"),
```
Comment (SemionPar) [inline plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java:410]: Some CASTs to char might land here, e.g.:

c_integer -> char(50)

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+                new CastTestCase("c_real", "double", Optional.of("c_double_precision")),
+                new CastTestCase("c_double_precision", "real", Optional.of("c_real")),
+                new CastTestCase("c_double_precision", "decimal(10,2)", Optional.of("c_decimal_10_2")),
+
+                new CastTestCase("c_varchar_unicode", "char(50)", Optional.of("c_char_50")),
+                new CastTestCase("c_nvarchar_unicode", "char(50)", Optional.of("c_char_50")),
+
+                new CastTestCase("c_timestamp", "date", Optional.of("c_date")),
+                new CastTestCase("c_timestamp", "time", Optional.of("c_time")),
+                new CastTestCase("c_date", "timestamp", Optional.of("c_timestamp")));
```
Comment (SemionPar) [inline plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java:398]: Some unsupported CASTs to varchar will land here, too, e.g.:
CAST(c_decimal_10_2 AS varchar(50))

**Review by SemionPar — COMMENTED** [review]
Skimmed

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftConnectorTest.java [hunk 4]
```diff
@@ -730,6 +730,35 @@ public void testJoinPushdownWithImplicitCast()
...
+    {
+        try (TestTable leftTable = new TestTable(
+                getQueryRunner()::execute,
+                "left_table_",
+                "(varchar_50 varchar(50))",
+                ImmutableList.of("('Value')", "('Valuea')"));
+                TestTable rightTable = new TestTable(
+                        getQueryRunner()::execute,
+                        "right_table_",
+                        "(varchar_50 varchar)",
```
Comment (SemionPar) [inline plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftConnectorTest.java:744]: *nit*: `varchar_50 varchar` -> `varchar_unb varchar` or just `c_var`?

File: plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementStringLpadFunction.java (hunk 1 above)
Comment (mayankvadariya) [reply plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementStringLpadFunction.java]: I had tried it earlier but the converted expression's type is [not considered rather column handles type is considered](https://github.com/trinodb/trino/blob/8fda4f04fd2cb823961393664d7f498d1acea581/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java#L465) and hence I had to pass empty column size in updatedTypeHandle and had to remove the [varchar precision constraint](https://github.com/trinodb/trino/pull/23490/commits/3c509210514835841ab018b3eb43ca339e1a94b7#diff-e7f2966e1e2e070e8643b5563c7ba62f7dc0652f76fbf446ef389b75ef4883fbL672-L675).
I left a java doc comment clarifying  it.

Comment (mayankvadariya) [reply plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementStringLpadFunction.java]: I had tried it earlier but the converted expression's type is [not considered rather column handles type is considered](https://github.com/trinodb/trino/blob/8fda4f04fd2cb823961393664d7f498d1acea581/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java#L465) and hence I passed an empty column size in `updatedTypeHandle` and had to remove the [varchar precision constraint](https://github.com/trinodb/trino/pull/23490/commits/3c509210514835841ab018b3eb43ca339e1a94b7#diff-e7f2966e1e2e070e8643b5563c7ba62f7dc0652f76fbf446ef389b75ef4883fbL672-L675). There is a [check](https://github.com/trinodb/trino/blob/7761b458a40f83a7833c1f85004f7b5feb43cbfa/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcRecordCursor.java#L91-L94) enforcing column handles's type to match with column mapping type.
I left a java doc comment clarifying  it.

**Comment by ebyhr** [comment]
/test-with-secrets sha=189caa3c735043a9f55c4b81b86ccb0e85217e62

File: plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementRedshiftLpad.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
+/*
```
Comment (ebyhr) [inline plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementRedshiftLpad.java:1]: Could you explain the motivation to support `lpad` function pusdhown in PR description?

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java (hunk 2 above)
Comment (mayankvadariya) [reply plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java:427]: thanks, added such cases in unsupported.

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java [hunk 6]
```diff
@@ -59,16 +59,25 @@ public void setupTable()
                 .addColumn("c_numeric_10_2", "numeric(10, 2)", asList(1.23, 2.67, null))
                 .addColumn("c_numeric_19_2", "numeric(19, 2)", asList(1.23, 2.67, null)) // Equal to REDSHIFT_DECIMAL_CUTOFF_PRECISION
                 .addColumn("c_numeric_30_2", "numeric(30, 2)", asList(1.23, 2.67, null))
+                .addColumn("c_char", "char", asList("'A'", "'B'", null))
+                .addColumn("c_character", "character", asList("'A'", "'B'", null))
```
Comment (krvikash) [inline plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java:63]: Can we have same data as other char and varchar columns.

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java [hunk 7]
```diff
@@ -124,16 +133,25 @@ public void setupTable()
                 .addColumn("c_numeric_10_2", "numeric(10, 2)", asList(1.23, 22.67, null))
                 .addColumn("c_numeric_19_2", "numeric(19, 2)", asList(1.23, 22.67, null)) // Equal to REDSHIFT_DECIMAL_CUTOFF_PRECISION
                 .addColumn("c_numeric_30_2", "numeric(30, 2)", asList(1.23, 22.67, null))
+                .addColumn("c_char", "char", asList("'A'", "'B'", null))
+                .addColumn("c_character", "character", asList("'A'", "'B'", null))
```
Comment (krvikash) [inline plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java:137]: ditto

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java [hunk 8]
```diff
@@ -267,6 +285,7 @@ public void testCastPushdownOverflow()
     protected List<CastTestCase> supportedCastTypePushdown()
     {
         return ImmutableList.of(
+
```
Comment (krvikash) [inline plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java:288]: remove blank line

File: plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementRedshiftLpad.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argument;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argumentCount;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.call;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.constant;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.functionName;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.type;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.variable;
+import static io.trino.spi.type.IntegerType.INTEGER;
+
+public class ImplementRedshiftLpad
```
Comment (krvikash) [inline plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementRedshiftLpad.java:47]: How about `RewritePad`?

File: plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementRedshiftLpad.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        Variable argument1 = captures.get(ARGUMENT1);
+        Constant argument2 = captures.get(ARGUMENT2);
+        Constant argument3 = captures.get(ARGUMENT3);
+
+        JdbcTypeHandle typeHandle = ((JdbcColumnHandle) context.getAssignment(argument1.getName())).getJdbcTypeHandle();
+
+        ParameterizedExpression translatedArgument1 = context.rewriteExpression(argument1).orElseThrow();
+        ParameterizedExpression translatedArgument2 = context.rewriteExpression(argument2).orElseThrow();
+        ParameterizedExpression translatedArgument3 = context.rewriteExpression(argument3).orElseThrow();
```
Comment (krvikash) [inline plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementRedshiftLpad.java:79]: Is there any case where `rewriteExpression` returns empty?

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftConnectorTest.java [hunk 11]
```diff
@@ -730,6 +730,35 @@ public void testJoinPushdownWithImplicitCast()
         }
     }
 
+    @Test
+    public void testLpadPushdown()
```
Comment (krvikash) [inline plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftConnectorTest.java:734]: can we add a simple test case for `lpad` on projection level as well? something like:

`SELECT lpad(l.c_varchar_50, 8, 'aaa') from table1`;

Comment (krvikash) [reply plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftConnectorTest.java:734]: Also the test coverage where length is greater and less than than column.

File: plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementRedshiftLpad.java
Comment (krvikash) [inline plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementRedshiftLpad.java:1]: I think `lpad` should be a separate PR.

**Review by krvikash — COMMENTED** [review]
@mayankvadariya can we rebase the PR with master and cherry-pick the latest changes https://github.com/trinodb/trino/pull/22951

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java (hunk 6 above)
Comment (mayankvadariya) [reply plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java:63]: Its because char/character without precision is char(1) https://docs.aws.amazon.com/redshift/latest/dg/r_Character_types.html#r_Character_types-char-or-character

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftConnectorTest.java (hunk 11 above)
Comment (mayankvadariya) [reply plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftConnectorTest.java:734]: sure, I can add one with greater than the column. We already have something simillar.

assertThat(query(session, "SELECT *, lpad(c_varchar_50, 20, 'aaa') FROM " + leftTable.getName()))
                    .isFullyPushedDown();

File: plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementRedshiftLpad.java
Comment (mayankvadariya) [reply plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/ImplementRedshiftLpad.java:1]: created https://github.com/trinodb/trino/pull/23718

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java (hunk 7 above)
Comment (mayankvadariya) [reply plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/BaseRedshiftCastPushdownTest.java]: Its because char/character without precision is char(1) https://docs.aws.amazon.com/redshift/latest/dg/r_Character_types.html#r_Character_types-char-or-character

**Comment by mayankvadariya** [comment]
Thanks for the review @krvikash.
Rebased on https://github.com/trinodb/trino/pull/22951 and added case-sensitive comparison.

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftCastPushdown.java [hunk 12]
```diff
@@ -491,6 +520,44 @@ protected List<CastTestCase> supportedCastTypePushdown()
...
+                new CastTestCase("c_varchar_10", "varchar(65535)", Optional.of("c_varchar_65535")),
+                new CastTestCase("c_varchar_10", "varchar(256)", Optional.of("c_nvarchar")),
+                new CastTestCase("c_varchar_10", "varchar(10)", Optional.of("c_nvarchar_10")),
+                new CastTestCase("c_varchar_10", "varchar(50)", Optional.of("c_nvarchar_50")),
+                new CastTestCase("c_varchar_10", "varchar(65535)", Optional.of("c_nvarchar_65535")),
+                new CastTestCase("c_varchar_10", "varchar(256)", Optional.of("c_text")),
+
+                new CastTestCase("c_varchar_unicode", "varchar(50)", Optional.of("c_varchar_50")),
+                new CastTestCase("c_nvarchar_unicode", "varchar(50)", Optional.of("c_varchar_50")),
+
```
Comment (krvikash) [inline plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftCastPushdown.java:581]: can we add test case for cast to `varchar` (unbounded varchar)?

Comment (krvikash) [reply plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftCastPushdown.java:581]: Something like https://github.com/mayankvadariya/trino/blob/41ba30e202dd5ab464d3d6bef7172e518d286db0/plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/TestOracleCastPushdown.java#L411-L445

**Comment by ebyhr** [comment]
#22951 is merged today. Could you rebase on master to resolve conflicts?

**Comment by mayankvadariya** [comment]
By mistake I closed this PR. Created https://github.com/trinodb/trino/pull/23808 as a replacement.
